sklearn/decomposition/fastica_.py

```diff
@@ -264,120 +266,37 @@ def my_g(x):
-
-    if whiten:
-        # Centering the columns (ie the variables)
-        X_mean = X.mean(axis=-1)
-        X -= X_mean[:, np.newaxis]
-
-        # Whitening and preprocessing by PCA
-        u, d, _ = linalg.svd(X, full_matrices=False)
-
-        del _
-        K = (u / d).T[:n_components]  # see (6.33) p.140
-        del u, d
-        X1 = np.dot(K, X)
-        # see (13.6) p.267 Here X1 is white and data
-        # in X has been projected onto a subspace by PCA
-        X1 *= np.sqrt(p)
-    else:
-        # X must be casted to floats to avoid typing issues with numpy
-        # 2.0 and the line below
-        X1 = as_float_array(X, copy=False)  # copy has been taken care of
 
-    if w_init is None:
-        w_init = np.asarray(random_state.normal(size=(n_components,
-                            n_components)), dtype=X1.dtype)
-
-    else:
-        w_init = np.asarray(w_init)
-        if w_init.shape != (n_components, n_components):
-            raise ValueError('w_init has invalid shape -- should be %(shape)s'
-                             % {'shape': (n_components, n_components)})
-
-    kwargs = {'tol': tol,
-              'g': g,
-              'fun_args': fun_args,
-              'max_iter': max_iter,
-              'w_init': w_init}
-
-    if algorithm == 'parallel':
-        W, n_iter = _ica_par(X1, **kwargs)
-    elif algorithm == 'deflation':
-        W, n_iter = _ica_def(X1, **kwargs)
-    else:
-        raise ValueError('Invalid algorithm: must be either `parallel` or'
-                         ' `deflation`.')
-    del X1
+    est = FastICA(n_components=n_components, algorithm=algorithm,
+                  whiten=whiten, fun=fun, fun_args=fun_args,
+                  max_iter=max_iter, tol=tol, w_init=w_init,
+                  random_state=random_state)
+    sources = est._fit(X, compute_sources=compute_sources)
 
     if whiten:
-        if compute_sources:
-            S = np.dot(np.dot(W, K), X).T
-        else:
-            S = None
         if return_X_mean:
             if return_n_iter:
-                return K, W, S, X_mean, n_iter
+                return (est.whitening_, est._unmixing, sources, est.mean_,
+                        est.n_iter_)
             else:
-                return K, W, S, X_mean
+                return est.whitening_, est._unmixing, sources, est.mean_
         else:
             if return_n_iter:
-                return K, W, S, n_iter
+                return est.whitening_, est._unmixing, sources, est.n_iter_
             else:
-                return K, W, S
+                return est.whitening_, est._unmixing, sources
 
     else:
-        if compute_sources:
-            S = np.dot(W, X).T
-        else:
-            S = None
         if return_X_mean:
             if return_n_iter:
-                return None, W, S, None, n_iter
+                return None, sources, None, est.n_iter_
             else:
-                return None, W, S, None
+                return None, sources, None
```

Comment:
This is changing the number of output arguments, and `est._unmixing` is missing?